File: mobile/src/services/PushService.ts
/**
 * Push Notification Service
 *
 * Stack:
 * - @react-native-firebase/messaging (FCM)
 * - Novu self-hosted (subscribers & credentials API)
 * - react-native-callkeep (CallKit iOS / ConnectionService Android)
 * - react-native-incall-manager (audio routing)
 */

import {Platform, AppState} from 'react-native';
import messaging from '@react-native-firebase/messaging';
import RNCallKeep from 'react-native-callkeep';
import {useCallStore} from '@/store/callStore';
import {handleOAuthCallback} from '@/hooks/useDeepLinks';
import {navigationRef} from '@/navigation/navigationRef';
import {appLog} from '@/utils/appLogger';
import {
  requestPushPermission,
  getPushToken,
  registerDeviceOnBackend,
  registerForPush,
} from '@/utils/push';

const CALL_CHANNEL_ID = 'vondic_calls';

class PushService {
  private pushToken: string | null = null;
  private callkeepConfigured = false;
  private initialized = false;
  private currentUserId: string | null = null;

  private tokenRefreshUnsubscribe: (() => void) | null = null;
  private messageUnsubscribe: (() => void) | null = null;
  private notificationOpenedUnsubscribe: (() => void) | null = null;

  async initialize(userId: string): Promise<void> {
    if (!userId) {
      appLog('PushService', 'Cannot initialize without userId');
      return;
    }

    if (this.initialized && this.currentUserId === userId) {
      if (this.pushToken) {
        await registerDeviceOnBackend(userId, this.pushToken);
      }
      return;
    }

    // Clean up if initializing with a different user
    if (this.initialized && this.currentUserId !== userId) {
      this.cleanup();
    }

    this.initialized = true;
    this.currentUserId = userId;

    const hasPermission = await requestPushPermission();
    appLog('PushService', 'Has notification permission: ' + hasPermission);

    // Настраиваем слушатели FCM в любом случае (в т.ч. для тихих data-only звонков)
    await this.setupFirebaseMessaging(userId);

    if (Platform.OS === 'ios') {
      await this.setupCallKeep();
    }
  }

  private async setupFirebaseMessaging(userId: string): Promise<void> {
    // 1. Get and register token
    try {
      const token = await getPushToken();
      if (token) {
        this.pushToken = token;
        await registerDeviceOnBackend(userId, token);
      }
    } catch (error) {
      appLog('PushService', 'Failed to get/register FCM token', error);
    }

    // 2. Listen for token refresh
    this.tokenRefreshUnsubscribe = messaging().onTokenRefresh(async (newToken: string) => {
      appLog('PushService', 'Token refreshed');
      this.pushToken = newToken;
      await registerDeviceOnBackend(userId, newToken);
    });

    // 4. Foreground message listener
    this.messageUnsubscribe = messaging().onMessage(async (remoteMessage) => {
      appLog('PushService', 'Notification received in foreground', remoteMessage);
      this.handleIncomingNotification(remoteMessage);
    });

    // 5. Notification opened listener (app in background/inactive)
    this.notificationOpenedUnsubscribe = messaging().onNotificationOpenedApp((remoteMessage) => {
      appLog('PushService', 'Notification opened from background', remoteMessage);
      this.handleNotificationOpen(remoteMessage);
    });

    // 6. Initial notification (app launched from closed state)
    messaging().getInitialNotification().then((remoteMessage) => {
      if (remoteMessage) {
        appLog('PushService', 'App opened from quit state via notification', remoteMessage);
        setTimeout(() => {
          this.handleNotificationOpen(remoteMessage);
        }, 1000);
      }
    });
  }

  private async setupCallKeep(): Promise<void> {
    const options = {
      ios: {
        appName: 'Вондик',
        includesCallsInRecents: false,
        supportsVideo: true,
        maximumCallGroups: '1',
        maximumCallsPerCallGroup: '1',
      },
      android: {
        alertTitle: 'Требуются разрешения',
        alertDescription: 'Приложению необходим доступ к учётным записям телефона',
        cancelButton: 'Отмена',
        okButton: 'ОК',
        additionalPermissions: [],
        foregroundService: {
          channelId: CALL_CHANNEL_ID,
          channelName: 'Звонки Вондик',
          notificationTitle: 'Идёт звонок Вондик',
          notificationIcon: 'ic_launcher',
        },
      },
    };

    try {
      RNCallKeep.setup(options);
      RNCallKeep.setAvailable(true);
      this.callkeepConfigured = true;

      RNCallKeep.addEventListener('answerCall', ({callUUID}) => {
        appLog('PushService', '[CallKeep] answerCall ' + callUUID);
        const state = useCallStore.getState();
        if (state.incomingCall) {
          state.acceptCall(state.incomingCall.socketId);
        }
        RNCallKeep.setCurrentCallActive(callUUID);
      });

      RNCallKeep.addEventListener('endCall', ({callUUID}) => {
        appLog('PushService', '[CallKeep] endCall ' + callUUID);
        const state = useCallStore.getState();
        if (state.incomingCall) {
          state.rejectCall(state.incomingCall.socketId);
        }
        RNCallKeep.endCall(callUUID);
      });

      RNCallKeep.addEventListener('didPerformSetMutedCallAction', ({muted}) => {
        const state = useCallStore.getState();
        state.setMuted(muted);
      });
    } catch (error) {
      appLog('PushService', '[CallKeep] Setup failed', error);
    }
  }

  private handleIncomingNotification(remoteMessage: any): void {
    const data = remoteMessage.data || {};
    const type = data.type as string;

    if (type === 'incoming_call') {
      this.showIncomingCallNotification(data);
    } else {
      appLog('PushService', 'Non-call notification received in foreground', remoteMessage);
    }
  }

  private handleNotificationOpen(remoteMessage: any): void {
    const data = remoteMessage.data || {};
    appLog('PushService', 'Notification opened with data', data);

    if (data.deeplink) {
      handleOAuthCallback(data.deeplink as string).catch(() => {});
      return;
    }

    // Handle incoming call notifications
    if (data.type === 'incoming_call') {
      const callerId = data.caller_user_id || data.caller_id;
      this.navigateWhenReady('Call', {
        targetUserId: data.group_id ? data.group_id : callerId,
        isIncoming: true,
        callerSocketId: data.caller_socket_id || callerId,
        isGroupCall: data.is_group === 'true' || !!data.group_id,
        callId: data.call_id,
        groupId: data.group_id,
      });
      return;
    }

    // Handle normal messages notifications
    const groupId = data.group_id;
    const channelId = data.channel_id;
    const senderId = data.sender_id || data.sender_user_id;

    if (groupId || channelId || senderId) {
      const type = groupId ? 'group' : channelId ? 'channel' : 'dm';
      const id = groupId || channelId || senderId;
      let name = remoteMessage.notification?.title || 'Чат';

      // Parse group name if title is like "Ivan (Group Name)"
      if (type === 'group' && name.includes('(') && name.endsWith(')')) {
        const idx = name.indexOf('(');
        name = name.slice(idx + 1, -1);
      }

      this.navigateWhenReady('Chat', {
        type,
        id,
        name,
      });
    }
  }

  private navigateWhenReady(screen: string, params: any) {
    if (navigationRef.isReady()) {
      navigationRef.navigate(screen as any, params);
      return;
    }

    let attempts = 0;
    const interval = setInterval(() => {
      attempts++;
      if (navigationRef.isReady()) {
        clearInterval(interval);
        navigationRef.navigate(screen as any, params);
      } else if (attempts > 50) { // 5 seconds max
        clearInterval(interval);
        appLog('PushService', 'Navigation ref not ready after 5 seconds');
      }
    }, 100);
  }

  private showIncomingCallNotification(data: Record<string, any>): void {
    const callerId = data.caller_user_id || data.caller_id || 'unknown';
    const callerName = data.caller_username || data.caller_name || 'Unknown';

    if (Platform.OS === 'ios' && this.callkeepConfigured) {
      const callUUID = `vondic-call-${callerId}-${Date.now()}`;
      RNCallKeep.displayIncomingCall(callUUID, String(callerId), callerName, 'generic', true);
    }

    const state = useCallStore.getState();
    state.setIncomingCall({
      socketId: data.caller_socket_id || callerId,
      userId: callerId,
      userName: callerName,
      avatarUrl: data.caller_avatar_url,
      status: 'ringing',
      startTime: new Date(),
    });
  }

  getToken(): string | null {
    return this.pushToken;
  }

  cleanup(): void {
    appLog('PushService', 'Cleaning up listeners...');
    if (this.tokenRefreshUnsubscribe) {
      this.tokenRefreshUnsubscribe();
      this.tokenRefreshUnsubscribe = null;
    }
    if (this.messageUnsubscribe) {
      this.messageUnsubscribe();
      this.messageUnsubscribe = null;
    }
    if (this.notificationOpenedUnsubscribe) {
      this.notificationOpenedUnsubscribe();
      this.notificationOpenedUnsubscribe = null;
    }
    this.initialized = false;
    this.pushToken = null;
    this.currentUserId = null;
  }
}

export const pushService = new PushService();

File: mobile_flutter/lib/core/utils/push_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter_callkit_incoming/entities/android_params.dart';
import 'package:flutter_callkit_incoming/entities/call_event.dart';
import 'package:flutter_callkit_incoming/entities/call_kit_params.dart';
import 'package:flutter_callkit_incoming/entities/ios_params.dart';
import 'package:flutter_callkit_incoming/flutter_callkit_incoming.dart';
import 'package:logger/logger.dart';
import '../network/api_client.dart';
import 'storage_service.dart';

// Top-level background message handler for FCM
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  final logger = Logger(printer: SimplePrinter(colors: true));
  logger.d('[PushService] Background message received: ${message.data}');
  
  final data = message.data;
  final type = data['type'] as String?;
  if (type == 'incoming_call') {
    final callerId = data['caller_user_id'] ?? data['caller_id'] ?? 'unknown';
    final callerName = data['caller_username'] ?? data['caller_name'] ?? 'Входящий звонок';
    final callId = data['call_id'] ?? '';
    final avatarUrl = data['caller_avatar_url'] ?? '';

    final params = CallKitParams(
      id: 'vondic-call-$callerId-${DateTime.now().millisecondsSinceEpoch}',
      nameCaller: callerName,
      appName: 'Вондик',
      avatar: avatarUrl,
      handle: callerId.toString(),
      type: 0, // audio call
      duration: 30000,
      extra: {'callId': callId, 'callerSocketId': data['caller_socket_id']},
      android: const AndroidParams(
        isCustomNotification: true,
        isShowLogo: false,
        ringtonePath: 'rington',
        backgroundColor: '#0f0f0f',
        actionColor: '#6c5ce7',
      ),
      ios: const IOSParams(
        iconName: 'AppIcon',
        handleType: 'generic',
        supportsVideo: true,
        maximumCallGroups: 1,
        maximumCallsPerCallGroup: 1,
        ringtonePath: 'rington.wav',
      ),
    );
    await FlutterCallkitIncoming.showCallkitIncoming(params);
  }
}

class PushService {
  final ApiClient _apiClient;
  final StorageService _storageService;
  final Logger _logger = Logger(printer: SimplePrinter(colors: true));

  String? _pushToken;
  bool _initialized = false;
  String? _currentUserId;

  // Callback to trigger call accept/decline inside CallStore/CallManager
  Function(String callerSocketId, bool accepted)? onCallKitEvent;
  // Callback to navigate to a screen
  Function(String screen, Map<String, dynamic> params)? onNavigate;

  PushService(this._apiClient, this._storageService);

  Future<void> initialize(String userId) async {
    if (userId.isEmpty) {
      _logger.w('[PushService] Cannot initialize without userId');
      return;
    }

    if (_initialized && _currentUserId == userId) {
      if (_pushToken != null) {
        await _registerDeviceOnBackend(userId, _pushToken!);
      }
      return;
    }

    _initialized = true;
    _currentUserId = userId;

    await _requestPushPermission();
    await _setupFirebaseMessaging(userId);
    await _setupCallKit();
  }

  Future<bool> _requestPushPermission() async {
    try {
      final messaging = FirebaseMessaging.instance;
      final settings = await messaging.requestPermission(
        alert: true,
        announcement: false,
        badge: true,
        carPlay: false,
        criticalAlert: false,
        provisional: false,
        sound: true,
      );
      final granted = settings.authorizationStatus == AuthorizationStatus.authorized ||
          settings.authorizationStatus == AuthorizationStatus.provisional;
      _logger.d('[PushService] Permission status: ${settings.authorizationStatus} | Granted: $granted');
      return granted;
    } catch (e) {
      _logger.e('[PushService] Error requesting permission: $e');
      return false;
    }
  }

  Future<void> _setupFirebaseMessaging(String userId) async {
    try {
      final messaging = FirebaseMessaging.instance;
      
      // 1. Get token
      final token = await messaging.getToken();
      if (token != null) {
        _pushToken = token;
        await _registerDeviceOnBackend(userId, token);
      }

      // 2. Listen to token refresh
      messaging.onTokenRefresh.listen((newToken) async {
        _logger.d('[PushService] Token refreshed');
        _pushToken = newToken;
        await _registerDeviceOnBackend(userId, newToken);
      });

      // 3. Foreground message listener
      FirebaseMessaging.onMessage.listen((RemoteMessage message) {
        _logger.d('[PushService] Message received in foreground: ${message.data}');
        _handleIncomingNotification(message);
      });

      // 4. Notification opened from background listener
      FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
        _logger.d('[PushService] Message opened from background: ${message.data}');
        _handleNotificationOpen(message);
      });

      // 5. Check if app launched from terminated state via push
      final initialMessage = await messaging.getInitialMessage();
      if (initialMessage != null) {
        _logger.d('[PushService] App opened from quit state via notification: ${initialMessage.data}');
        Future.delayed(const Duration(seconds: 1), () {
          _handleNotificationOpen(initialMessage);
        });
      }
    } catch (e) {
      _logger.e('[PushService] Error setting up FCM listeners: $e');
    }
  }

  Future<void> _registerDeviceOnBackend(String userId, String token) async {
    try {
      _logger.d('[PushService] Registering device on backend...');
      final response = await _apiClient.post('/devices/register', data: {
        'token': token,
        'platform': Platform.isIOS ? 'ios' : 'android',
        'device_type': 'mobile',
      });
      
      final data = response.data;
      if (response.statusCode == 200 && (data['success'] == true || data['ok'] == true)) {
        _logger.d('[PushService] Device successfully registered on backend');
      } else {
        _logger.e('[PushService] Backend failed to register device: $data');
      }
    } catch (e) {
      _logger.e('[PushService] Exception registering device: $e');
    }
  }

  Future<void> _setupCallKit() async {
    try {
      FlutterCallkitIncoming.onEvent.listen((CallEvent? event) {
        if (event == null) return;
        
        if (event is CallEventActionCallAccept) {
          final extra = event.callKitParams.extra;
          final callerSocketId = extra?['callerSocketId'] as String? ?? '';
          _logger.d('[CallKit] Answered call from socket: $callerSocketId');
          if (onCallKitEvent != null && callerSocketId.isNotEmpty) {
            onCallKitEvent!(callerSocketId, true);
          }
        } else if (event is CallEventActionCallDecline) {
          final extra = event.callKitParams.extra;
          final callerSocketId = extra?['callerSocketId'] as String? ?? '';
          _logger.d('[CallKit] Declined call from socket: $callerSocketId');
          if (onCallKitEvent != null && callerSocketId.isNotEmpty) {
            onCallKitEvent!(callerSocketId, false);
          }
        }
      });
    } catch (e) {
      _logger.e('[PushService] Setup CallKit failed: $e');
    }
  }

  void _handleIncomingNotification(RemoteMessage message) {
    final data = message.data;
    final type = data['type'] as String?;

    if (type == 'incoming_call') {
      _showIncomingCallNotification(data);
    }
  }

  void _handleNotificationOpen(RemoteMessage message) {
    final data = message.data;
    if (data.isEmpty) return;

    if (data['type'] == 'incoming_call') {
      final callerId = data['caller_user_id'] ?? data['caller_id'];
      if (onNavigate != null) {
        onNavigate!('Call', {
          'targetUserId': data['group_id'] != null ? data['group_id'] : callerId,
          'isIncoming': true,
          'callerSocketId': data['caller_socket_id'] ?? callerId,
          'isGroupCall': data['is_group'] == 'true' || data['group_id'] != null,
          'callId': data['call_id'],
          'groupId': data['group_id'],
        });
      }
      return;
    }

    final groupId = data['group_id'] as String?;
    final channelId = data['channel_id'] as String?;
    final senderId = data['sender_id'] ?? data['sender_user_id'];

    if (groupId != null || channelId != null || senderId != null) {
      final type = groupId != null ? 'group' : channelId != null ? 'channel' : 'dm';
      final id = groupId ?? channelId ?? senderId.toString();
      final title = message.notification?.title ?? 'Сообщение';

      if (onNavigate != null) {
        onNavigate!('Chat', {
          'type': type,
          'id': id,
          'name': title,
        });
      }
    }
  }

  Future<void> _showIncomingCallNotification(Map<String, dynamic> data) async {
    final callerId = data['caller_user_id'] ?? data['caller_id'] ?? 'unknown';
    final callerName = data['caller_username'] ?? data['caller_name'] ?? 'Входящий звонок';
    final callId = data['call_id'] ?? '';
    final avatarUrl = data['caller_avatar_url'] ?? '';

    final callUUID = 'vondic-call-$callerId-${DateTime.now().millisecondsSinceEpoch}';
    final params = CallKitParams(
      id: callUUID,
      nameCaller: callerName,
      appName: 'Вондик',
      avatar: avatarUrl,
      handle: callerId.toString(),
      type: 0, // audio call
      duration: 30000,
      extra: {'callId': callId, 'callerSocketId': data['caller_socket_id']},
      android: const AndroidParams(
        isCustomNotification: true,
        isShowLogo: false,
        ringtonePath: 'rington',
        backgroundColor: '#0f0f0f',
        actionColor: '#6c5ce7',
      ),
      ios: const IOSParams(
        iconName: 'AppIcon',
        handleType: 'generic',
        supportsVideo: true,
        maximumCallGroups: 1,
        maximumCallsPerCallGroup: 1,
        ringtonePath: 'rington.wav',
      ),
    );
    await FlutterCallkitIncoming.showCallkitIncoming(params);
  }
}
